Terminal - reworked presentation to show two-liner of Terminal and toggle to maximization

pskowronek opened the pull request:
Terminal - reworked presentation to show two-liner of Terminal and toggle to maximization.

As discussed here: https://github.com/mucommander/mucommander/issues/977#issuecomment-1656641763

MuC initial state:
![obraz](https://github.com/mucommander/mucommander/assets/360665/59a95b8d-9934-4bcc-aac5-4e9e3cc425e7)

After clicking Up on divider (or on divider, or shortcut):

![obraz](https://github.com/mucommander/mucommander/assets/360665/2e758e9e-d575-4509-b8ff-2648b202f78b)

After clicking Down on divider (or on divider, or shortcut):

![obraz](https://github.com/mucommander/mucommander/assets/360665/d67e060c-d962-4c51-b8de-ec397901b550)

After clicking one more time Down on divider (or when terminal is `exit`ed)

![obraz](https://github.com/mucommander/mucommander/assets/360665/9a531bc8-9547-4e71-99a4-281033d9b8cf)

One can move/drag divider manually then its position is being remember as 'minimized' state (to have more lines than 2), however it divider reaches 2/3rds in height then it is being assumed as 'maximized' state and it will be ignored when hiding.

pskowronek commented:
> Another problem: In the state where the terminal and both file panes are shown, switching between left and right pane does not trigger a `cd` command in the terminal.

PR updated to address this one

pskowronek commented:
> Using FlatLaf Light, in the initial state, the divider doesn't look good. And even worse, clicking Up has no effect.

As it comes to 'look good' - what do you mean exactly, @Andreas0602 ?
It looks the same as the existing horizontal splitter (i.e. vertical bar/divider):
![obraz](https://github.com/mucommander/mucommander/assets/360665/a8d9152a-8c7c-4d8e-b30b-e722da8d8f45)

On my mac Up button works - it is not so precise as for `Mac OS X` theme, but it works if you aim right. Have you restarted muC after changing the theme or not?

pskowronek commented:
> There's definitely a problem in the initial state. The Up arrow doesn't work in that state.

can you check logs? For me it works, when Terminal has to initialize for the first time, it comes around 1-2s (and I've got kinda old macbook).

pskowronek commented:
> I mean that the divider is not clearly recognizable. At least that's true for the initial state (after starting muCommander) and in the 'maximized terminal' state.

No idea how to make it more visible or obvious, however, if you hang mouse on that divider a tooltip should show up to give you a hint what it does. If somebody knows that Terminal functionality exists then one will just press F12 or sth. 

We might think if we would like to add an option to preferences to start terminal when muC is started and show it in 2-liner mode.

pskowronek commented on build.gradle:
```diff
@@ -139,7 +139,11 @@ runOsgi {
+            --add-opens java.base/java.net=ALL-UNNAMED \
+            --add-exports java.desktop/com.apple.eawt=ALL-UNNAMED \
+            --add-exports java.desktop/com.apple.laf=ALL-UNNAMED  \
+            --add-exports java.desktop/com.apple.eio=ALL-UNNAMED'
```
just split into separate lines to make it easier to maintain

pskowronek commented on build.gradle:
```diff
@@ -340,7 +344,20 @@ task macAppImage(dependsOn: 'createBundlesDir', type: Exec) {
+                        --add-exports java.desktop/com.apple.eio=ALL-UNNAMED \
+                        --add-exports java.desktop/com.apple.eawt=ALL-UNNAMED \
+                        --add-opens java.desktop/javax.swing.plaf.basic=ALL-UNNAMED \
+                        --add-opens java.base/java.io=ALL-UNNAMED \
```
added this line (with javax.swing.plaf.basic). Should I also add it to other places? Like the one above or in mucommander.sh?

Is there some similar place for other platforms like unix/windows? @ahadas

ahadas replied:
yeah, it should also be added to `java-options` in `winAppImage` and `linuxAppImage`
`mucommander.sh` is used in tgz and portable packages now and I don't regularly use them - seems like it missed few things.. but anyway, as for this change - better to add it there as well

pskowronek replied:
done

pskowronek commented on package/unix/mucommander.sh:
```diff
@@ -35,4 +35,9 @@ fi
+      --add-exports java.desktop/com.apple.eawt=ALL-UNNAMED \
+      --add-exports java.desktop/com.apple.laf=ALL-UNNAMED  \
+      --add-exports java.desktop/com.apple.eio=ALL-UNNAMED \
+      -Djava.library.path=/usr/local/lib -cp mucommander-@MU_VERSION@.jar com.mucommander.main.muCommander $@
```
just split to separate lines to make easier to maintain

ahadas commented on mucommander-core/src/main/java/com/mucommander/ui/main/MainFrame.java:
```diff
@@ -206,17 +209,34 @@ public Insets getInsets() {
-        splitPane.disableAccessibilityShortcuts();
+        foldersSplitPane.disableAccessibilityShortcuts();
 
+        verticalSplitPane = new JSplitPane( JSplitPane.VERTICAL_SPLIT, true) {
```
```suggestion
        verticalSplitPane = new JSplitPane(JSplitPane.VERTICAL_SPLIT, true) {
```

pskowronek replied:
done

ahadas commented on mucommander-core/src/main/java/com/mucommander/ui/main/MainFrame.java:
```diff
@@ -206,17 +209,34 @@ public Insets getInsets() {
+        };
+        verticalSplitPane.setBorder(null);
+        verticalSplitPane.setOneTouchExpandable(true);
+        // TODO?
```
I see the comment above:
```
// Disable all the JSplitPane accessibility shortcuts that are registered by default, as some of them
// conflict with default mucommander action shortcuts (e.g. F6 and F8)
```
which seems to hold to the verticalSplitPane as well

pskowronek replied:
done

ahadas commented on mucommander-core/src/main/java/com/mucommander/ui/main/MainFrame.java:
```diff
@@ -206,17 +209,34 @@ public Insets getInsets() {
+        verticalSplitPane.setBottomComponent(null);
+
+        JLayeredPane layeredPane = new JLayeredPane();
+        layeredPane.setLayout(new BorderLayout() );
```
```suggestion
        layeredPane.setLayout(new BorderLayout());
```

pskowronek replied:
done

ahadas commented on mucommander-core/src/main/java/com/mucommander/ui/terminal/TerminalIntegration.java:
```diff
@@ -0,0 +1,304 @@
+                    });
+                    cwd = newCwd;
+                    terminal.getTerminalPanel().addCustomKeyListener(termCloseKeyHandler());
+                    // TODO do this better? For now 2 lines ~height + 20%
```
TODO for myself - comment after I play with that

ahadas replied:
after playing with it, the height looks good to me
but I'd suggest to not to minimize/hide the terminal when mucommander starts, it would be better to show it, in its minimal size, so users would notice it
I also miss the option to switch back from forth from the terminal to the file table using the keyboard - there's no shortcut for this, right?

pskowronek replied:
I think that since Terminal uses resources, I'm not sure if it is really welcomed change? We may add an options to Preferences to control that?

pskowronek replied:
> I also miss the option to switch back from forth from the terminal to the file table using the keyboard - there's no shortcut for this, right?

You mean when both are visible you would like to switch active panels? (Otherwise F12)

ahadas replied:
no, not to switch active panels but rather something like the following scenario:
1. switch to the terminal
2. make the terminal a big larger
3. start building something (e.g., `mvn install`)
4. switch back to the folder table and navigate to a sub folder there while keep seeing the terminal and the progress of the build in the terminal below

pskowronek replied:
this is what I meant. For that there's no shortcut - we may think about it outside of this PR I think

pskowronek replied:
> I think that since Terminal uses resources, I'm not sure if it is really welcomed change? We may add an option to Preferences to control that?

And I think there are two types of users - geeks that are familiar with terminal and civilians that might be scared by "black window" - those who come from Windows' Total Commander. So, for "geeks" I would just add preferences panel to launch terminal in minimized 2-liner form (by default set to `false`).  Tell me what you think and I will add it.

ahadas replied:
> this is what I meant. For that there's no shortcut - we may think about it outside of this PR I think

ah ok, I thought you meant the two file table panels, didn't realize you also referred to the terminal as a panel

> And I think there are two types of users - geeks that are familiar with terminal and civilians that might be scared by "black window" - those who came from Windows' Total Commander. So, for "geeks" I would just add preferences panel to launch terminal in minimized 2-liner form (by default set to false). Tell me what you think and I will add it.

I think it makes sense to make configurable, I'm not sure if disabling it by default make more sense than enabling it by default - if we hide the terminal by default, it's less visible for users that may be interested in using an integrated terminal. how about showing the terminal by default and add an option, like `customize command bar`, that can be used to set that preference you proposed to `false`?

pskowronek replied:
OK, however `customize command bar` even for me is kinda ambiguous, we need a better wording for that

ahadas replied:
sure, we can find better wording
this PR introduce an improvement even without this part, would you like to include it here or do it separately?

pskowronek replied:
> sure, we can find better wording this PR introduce an improvement even without this part, would you like to include it here or do it separately?

I will provide a separate PR for that. In the meantime, I've rebased this PR and added feature description to readme.

ahadas reviewed (commented):
I see @Andreas0602 played with it already and couple of issues were sorted out (good!), I'll follow up on this PR after playing with it myself, but code-wise it looks good!

pskowronek commented:
btw, should we change functionality of this toolbar icon? 
![obraz](https://github.com/mucommander/mucommander/assets/360665/90edecd5-b9b5-480a-882b-9437223e6290)

to show this integrated terminal? 
Should we keep "Run command" i.e. "external" Terminal window functionality?

ahadas commented:
> btw, should we change functionality of this toolbar icon? ![obraz](https://user-images.githubusercontent.com/360665/260148405-90edecd5-b9b5-480a-882b-9437223e6290.png)
> 
> to show this integrated terminal? 

+1 for changing the functionality of that toolbar icon

> Should we keep "Run command" i.e. "external" Terminal window functionality?

I don't think so - now that we use the same things behind the scenes for both things and the integrated terminal works well, I don't see a reason to keep the 'run command' dialog

pskowronek commented:
PR updated after PR review

pskowronek commented on mucommander-translator/src/main/resources/dictionary_en.properties:
```diff
@@ -241,7 +241,7 @@ CustomizeCommandBar.label = Customize command bar
 ToggleStatusBar.hide = Hide status bar
 ToggleShowFoldersFirst.label = Show folders first
-ToggleTerminal.label = Show terminal
+ToggleTerminal.show = Show terminal
```
done

pskowronek commented:
> btw, should we change functionality of this toolbar icon? ![obraz](https://user-images.githubusercontent.com/360665/260148405-90edecd5-b9b5-480a-882b-9437223e6290.png)
> 
> to show this integrated terminal? Should we keep "Run command" i.e. "external" Terminal window functionality?

removed and replaced with integrated Terminal

pskowronek commented:
> @pskowronek and @ahadas, in the initial state (after startup), clicking the Up arrow has no effect, and dragging the divider upwards has no effect either. The logs don't contain an error.

https://github.com/mucommander/mucommander/pull/1043

pskowronek commented:
> In the initial state (after startup), clicking the Up arrow in #1043 is okay now, but dragging the divider upwards still has no effect.

This is by design since Terminal underneath is not initialized. We may think about it along PR for Preferences and command line visibility.

ahadas commented:
Version 1.4.0 that includes a fix for this issue [has been released](https://github.com/mucommander/mucommander/releases/tag/1.4.0-1)
